src/liballoc/vec.rs

```diff
@@ -2000,7 +1999,137 @@ impl<T, I> SpecExtend<T, I> for Vec<T>
         <Vec<T> as SpecExtend<T, I>>::spec_extend(&mut vector, iterator);
         vector
     }
+}
+
+struct InPlaceDrop<T> {
+    inner: *mut T,
+    dst: *mut T,
+    did_panic: bool,
+}
+
+impl<T> InPlaceDrop<T> {
+    unsafe fn len(&self) -> usize {
+        self.dst.offset_from(self.inner) as usize
+    }
+}
+
+impl<T> Drop for InPlaceDrop<T> {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe {
+            if self.did_panic {
+                ptr::drop_in_place(slice::from_raw_parts_mut(self.inner, self.len()) as *mut _);
+            }
+        }
+    }
+
+}
+
+fn from_into_iter_source<T, I>(mut iterator: I) -> Vec<T>
+    where I: Iterator<Item=T> + InPlaceIterable + SourceIter<Source=IntoIter<T>>
+{
+    let src_buf = iterator.as_inner().buf.as_ptr();
+    let src_end = iterator.as_inner().end;
+    let dst = src_buf;
+
+    let dst = if mem::needs_drop::<T>() {
+        // special-case drop handling since it prevents vectorization
+        let mut sink = InPlaceDrop { inner: src_buf, dst, did_panic: true };
+        let _ = iterator.try_for_each::<_, Result<_, !>>(|item| {
```

Comment:
Why is this using try_for_each if it's infallible?

Reply:
`for_each(self)` vs. `try_for_each(&mut self)`. I need to keep `iterator` alive.